### PR TITLE
feat(amazon): ASG support for features enabled by launch templates

### PR DESCRIPTION
**Pull request by caseyhebebrand**
1. Support for cloning an ASG that is launched with a launch template. 
2. Based on aws provider settings, allow opt-in to IPv6 or IMDSv2 when creating a server group.
Screen Shot 2020-06-03 at 3 05 19 PM

### CODE REVIEW

File: settings.js [hunk 1]
```diff
@@ -2,34 +2,34 @@
 
 // Use environment variables when developing locally via 'yarn start', i.e.:
 // API_HOST=https://gate.spinnaker.mycompany.com yarn start
-var apiHost = process.env.API_HOST || 'http://localhost:8084';
-var atlasWebComponentsUrl = process.env.ATLAS_WEB_COMPONENTS_URL;
-var authEndpoint = process.env.AUTH_ENDPOINT || apiHost + '/auth/user';
-var authEnabled = process.env.AUTH_ENABLED === 'false' ? false : true;
-var bakeryDetailUrl =
+const apiHost = process.env.API_HOST || 'http://localhost:8084';
```
Comment (caseyhebebrand) [inline settings.js:5]: Linter magic

File: app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js [hunk 2]
```diff
@@ -333,13 +341,45 @@ angular
...
+
           if (mode === 'clone' && serverGroup.image && serverGroup.image.name) {
             command.amiName = serverGroup.image.name;
           }
 
           if (serverGroup.launchConfig && serverGroup.launchConfig.securityGroups.length) {
             command.securityGroups = serverGroup.launchConfig.securityGroups;
           }
+
+          if (serverGroup.launchTemplate && serverGroup.launchTemplate.launchTemplateData.securityGroups.length) {
```
Comment (jeyrschabu) [inline app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js:379]: The backend is setting security groups on network interfaces. (Ran into issues with cloneServerGroups when the groups aren’t set at the network interface level.

File: app/scripts/modules/amazon/src/serverGroup/configure/wizard/pages/advancedSettings/ServerGroupAdvancedSettingsCommon.tsx [hunk 3]
```diff
@@ -255,41 +273,58 @@ export class ServerGroupAdvancedSettingsCommon extends React.Component<IServerGr
...
-            </label>
+            <b>{`Associate ${asgSettings?.enableIPv6 ? 'IPv6' : 'Public IPv4'} Address`}</b>
           </div>
+          {asgSettings?.enableIPv6 && (
+            <div className="col-md-6 checkbox">
+              <label>
+                <input
+                  type="checkbox"
+                  checked={values.associatePublicIpAddress === true}
+                  onChange={e => setFieldValue('associatePublicIpAddress', e.target.checked)}
```
Comment (jeyrschabu) [inline app/scripts/modules/amazon/src/serverGroup/configure/wizard/pages/advancedSettings/ServerGroupAdvancedSettingsCommon.tsx:284]: associatePublicIpAddress controls whether or not to assign a public ipv4 address on both a launch config and template. 
I need to look at how to auto assign IPv6 with launch templates.

File: app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js (hunk 2 above)
Comment (caseyhebebrand) [reply app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js:379]: Ok, is it problematic to set them here too or should I just set them on both `securityGroups` and `networkInterfaces`?

File: app/scripts/modules/amazon/src/serverGroup/configure/wizard/pages/advancedSettings/ServerGroupAdvancedSettingsCommon.tsx (hunk 3 above)
Comment (caseyhebebrand) [reply app/scripts/modules/amazon/src/serverGroup/configure/wizard/pages/advancedSettings/ServerGroupAdvancedSettingsCommon.tsx:284]: When launch templates are enabled, CNE wants assigning a public IP address to be IPv6 only. Any dual-stacked instance will have a private IPv4 and an IPv6. We can either re-use this flag, and perform a different operation in clouddriver when using launch templates or we can use a new flag `assignIPv6`.  Let me know what makes sense. The change on my end will be pretty nominal.

Comment (jeyrschabu) [reply app/scripts/modules/amazon/src/serverGroup/configure/wizard/pages/advancedSettings/ServerGroupAdvancedSettingsCommon.tsx:284]: Let’s reuse the same flag.

File: app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js (hunk 2 above)
Comment (jeyrschabu) [reply app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js:379]: Just securityGroups should be fine. Groups can be found here `serverGroup.launchTemplate.launchTemplateData.networkInterfaces*.groups`.  Before you change anything, let me try a couple of things.

**Comment by christopherthielen** [comment]
lgtm!

File: app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js [hunk 4]
```diff
@@ -333,13 +341,45 @@ angular
             command.viewState.imageId = serverGroup.launchConfig.imageId;
           }
 
+          if (serverGroup.launchTemplate) {
+            const { launchTemplateData } = serverGroup.launchTemplate;
+            const maxPrice =
+              launchTemplateData.metadataOptions &&
+              launchTemplateData.metadataOptions.spotOptions &&
+              launchTemplateData.metadataOptions.spotOptions.maxPrice;
```
[the remark this answers is not in the document]
Comment (caseyhebebrand) [reply app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js:349]: Optional chaining was throwing build errors, but let's see how much of a pain it is to make this a `.ts`

[the remark this answers is not in the document]
Comment (christopherthielen) [reply app/scripts/modules/amazon/src/serverGroup/configure/serverGroupCommandBuilder.service.js]: oh I didn't even notice this was .js.  I wouldn't worry about it then

**Pull request merged**